### PR TITLE
Implement sort/sorted with key.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

File: numba/targets/listobj.py [hunk 1]
```diff
@@ -1031,57 +1032,85 @@ def list_reverse_impl(lst):
...
+        raise errors.TypingError(msg)
+
+
+@overload_method(types.List, "sort")
+def ol_list_sort(lst, key=None, reverse=False):
+
+    _sort_check_key(key)
+    _sort_check_reverse(reverse)
+
+    if (isinstance(key, (types.NoneType, types.Omitted)) or key is None):
```
Comment (sklam) [inline numba/targets/listobj.py]: use `cgutils.is_nonelike()` instead

File: numba/tests/test_sort.py [hunk 2]
```diff
@@ -923,5 +927,149 @@ def test_argsort_stable(self):
...
+
+        a = [3, 1, 4, 1, 5, 9]
+
+        @njit
+        def external_key(z):
+            return 1. / z
+
+        @njit
+        def foo(x, key=None):
+            return sorted(x[:], key=key), x[:].sort(key=key)
```
Comment (sklam) [inline numba/tests/test_sort.py]: `.sort` is returning `None`. it's not testing the sorted result.

File: numba/tests/test_sort.py [hunk 3]
```diff
@@ -923,5 +927,149 @@ def test_argsort_stable(self):
...
+
+    def test_02(self):
+
+        a = [3, 1, 4, 1, 5, 9]
+
+        @njit
+        def foo(x):
+            def closure_key(z):
+                return 1. / z
+            return sorted(x[:], key=closure_key), x[:].sort(key=closure_key)
```
Comment (sklam) [inline numba/tests/test_sort.py]: `.sort` is returning `None`

File: numba/tests/test_sort.py [hunk 4]
```diff
@@ -923,5 +927,149 @@ def test_argsort_stable(self):
...
+
+    def test_03(self):
+
+        a = [3, 1, 4, 1, 5, 9]
+
+        def gen(compiler):
+
+            @compiler
+            def bar(x, func):
+                return sorted(x[:], key=func), x[:].sort(key=func)
```
Comment (sklam) [inline numba/tests/test_sort.py]: `.sort` is returning `None`

File: numba/tests/test_sort.py [hunk 5]
```diff
@@ -923,5 +927,149 @@ def test_argsort_stable(self):
...
+
+        a = ['a','b','B','b','C','A']
+
+        @njit
+        def external_key(z):
+            return z.upper()
+
+        @njit
+        def foo(x, key=None):
+            return sorted(x[:], key=key), x[:].sort(key=key)
```
Comment (sklam) [inline numba/tests/test_sort.py]: `.sort` is returning `None`

File: numba/tests/test_sort.py [hunk 6]
```diff
@@ -923,5 +927,149 @@ def test_argsort_stable(self):
...
+        a = ['a','b','B','b','C','A']
+
+        @njit
+        def external_key(z):
+            return z.upper()
+
+        @njit
+        def foo(x, key=None, reverse=False):
+            return (sorted(x[:], key=key, reverse=reverse),
+                    x[:].sort(key=key, reverse=reverse))
```
Comment (sklam) [inline numba/tests/test_sort.py]: `.sort` is returning `None`

File: numba/tests/test_sort.py [hunk 7]
```diff
@@ -923,5 +927,149 @@ def test_argsort_stable(self):
...
+        @njit
+        def foo(x, predicate):
+            if predicate:
+                def closure_key(z):
+                    return 1. / z
+            else:
+                closure_key = None
+
+            return (sorted(x[:], key=closure_key),
+                    x[:].sort(key=closure_key))
```
Comment (sklam) [inline numba/tests/test_sort.py]: `.sort` is returning `None`

File: numba/tests/test_sort.py [hunk 8]
```diff
@@ -923,5 +927,149 @@ def test_argsort_stable(self):
...
+
+    def test_exceptions_sorted(self):
+
+        @njit
+        def foo_sorted(x, key=None, reverse=False):
+            return sorted(x[:], key=key, reverse=reverse)
+
+        @njit
+        def foo_sort(x, key=None, reverse=False):
+            return x[:].sort(key=key, reverse=reverse)
```
Comment (sklam) [inline numba/tests/test_sort.py]: `.sort` is returning `None`

File: numba/tests/test_sort.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/tests/test_sort.py]: fair point :)

File: numba/targets/listobj.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/targets/listobj.py]: Why is this in `cgutils`? It's typing not codegen?

Comment (sklam) [reply numba/targets/listobj.py]: good point. there's no typing util either. will need to sort that out later.

**Pull request merged**